From 161575434a3c32c81b32531839458f87f623bc5f Mon Sep 17 00:00:00 2001
From: Salvo Giangreco <giangrecosalvo9@gmail.com>
Date: Tue, 26 Aug 2025 00:06:07 +0200
Subject: [PATCH] Allow custom platform signature

Based off: https://github.com/EverestOS-AOSP/frameworks_base/commit/9d302efb52770dfb85e19c51ad31f781c7844ca3

UN1CA: retargeted using the available `.orig`/`.rej` smali artifacts and cross-checked with `unica/mods/blocklist/services.jar/0001-Allow-custom-PackageBlockListPolicy.patch`.

UN1CA: this patch is not complete! Please check unica/patches/signature/customize.sh for more info.
---
 .../android/server/pm/InstallPackageHelper.smali   |  72 +++++++-
 .../server/pm/PackageManagerServiceUtils.smali     |  10 ++
 .../com/android/server/pm/ScanPackageUtils.smali   | 184 +++++++++++++++------
 3 files changed, 205 insertions(+), 61 deletions(-)

diff --git a/smali_classes2/com/android/server/pm/InstallPackageHelper.smali b/smali_classes2/com/android/server/pm/InstallPackageHelper.smali
index fed2903d..0afae29a 100644
--- a/smali_classes2/com/android/server/pm/InstallPackageHelper.smali
+++ b/smali_classes2/com/android/server/pm/InstallPackageHelper.smali
@@ -12,6 +12,8 @@
 
 .field public final mContext:Landroid/content/Context;
 
+.field public final mCustomPlatformSignatures:[Landroid/content/pm/Signature;
+
 .field public final mDeletePackageHelper:Lcom/android/server/pm/DeletePackageHelper;
 
 .field public final mIncrementalManager:Landroid/os/incremental/IncrementalManager;
@@ -149,6 +151,18 @@
 
     iput-object p1, p0, Lcom/android/server/pm/InstallPackageHelper;->mInstallingPackages:Ljava/util/Set;
 
+    const-string p1, "CONFIG_CUSTOM_PLATFORM_SIGNATURE"
+
+    filled-new-array {p1}, [Ljava/lang/String;
+
+    move-result-object p1
+
+    invoke-static {p1}, Lcom/android/server/pm/InstallPackageHelper;->createSignatures([Ljava/lang/String;)[Landroid/content/pm/Signature;
+
+    move-result-object p1
+
+    iput-object p1, p0, Lcom/android/server/pm/InstallPackageHelper;->mCustomPlatformSignatures:[Landroid/content/pm/Signature;
+
     sget-boolean p1, Lcom/samsung/android/rune/PMRune;->PM_BADGE_ON_MONETIZED_APP_SUPPORTED:Z
 
     if-eqz p1, :cond_0
@@ -283,6 +297,34 @@
     return p0
 .end method
 
+.method public static createSignatures([Ljava/lang/String;)[Landroid/content/pm/Signature;
+    .locals 5
+
+    array-length v0, p0
+
+    new-array v1, v0, [Landroid/content/pm/Signature;
+
+    const/4 v2, 0x0
+
+    :goto_0
+    if-ge v2, v0, :cond_0
+
+    new-instance v3, Landroid/content/pm/Signature;
+
+    aget-object v4, p0, v2
+
+    invoke-direct {v3, v4}, Landroid/content/pm/Signature;-><init>(Ljava/lang/String;)V
+
+    aput-object v3, v1, v2
+
+    add-int/lit8 v2, v2, 0x1
+
+    goto :goto_0
+
+    :cond_0
+    return-object v1
+.end method
+
 .method public static isAdminApplication(Lcom/android/server/pm/pkg/AndroidPackage;)Z
     .locals 5
 
@@ -22501,7 +22543,9 @@
 
     const/4 v9, 0x1
 
-    invoke-static {v1, v10, v2, v9}, Lcom/android/server/pm/ScanPackageUtils;->applyPolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;ILcom/android/server/pm/pkg/AndroidPackage;Z)V
+    iget-object v5, v15, Lcom/android/server/pm/InstallPackageHelper;->mCustomPlatformSignatures:[Landroid/content/pm/Signature;
+
+    invoke-static {v1, v10, v2, v9, v5}, Lcom/android/server/pm/ScanPackageUtils;->applyPolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;ILcom/android/server/pm/pkg/AndroidPackage;Z[Landroid/content/pm/Signature;)V
 
     iget-object v2, v15, Lcom/android/server/pm/InstallPackageHelper;->mPm:Lcom/android/server/pm/PackageManagerService;
 
@@ -23042,15 +23086,27 @@
 
     invoke-virtual {v2, v1}, Lcom/android/server/pm/PackageManagerService;->getSettingsVersionForPackage(Lcom/android/server/pm/pkg/AndroidPackage;)Lcom/android/server/pm/Settings$VersionInfo;
 
-    move-result-object v2
+    move-result-object v7
 
-    iget-object v4, v15, Lcom/android/server/pm/InstallPackageHelper;->mPm:Lcom/android/server/pm/PackageManagerService;
+    move v4, v3
 
-    iget-boolean v5, v4, Lcom/android/server/pm/PackageManagerService;->mIsPreNMR1Upgrade:Z
+    move-object v3, v7
 
-    move/from16 v4, v16
+    iget-object v2, v15, Lcom/android/server/pm/InstallPackageHelper;->mPm:Lcom/android/server/pm/PackageManagerService;
+
+    iget-object v2, v2, Lcom/android/server/pm/PackageManagerService;->mPlatformPackage:Lcom/android/server/pm/pkg/AndroidPackage;
+
+    iget-object v5, v15, Lcom/android/server/pm/InstallPackageHelper;->mPm:Lcom/android/server/pm/PackageManagerService;
+
+    iget-boolean v6, v5, Lcom/android/server/pm/PackageManagerService;->mIsPreNMR1Upgrade:Z
 
-    invoke-static/range {v0 .. v5}, Lcom/android/server/pm/ScanPackageUtils;->collectCertificatesLI(Lcom/android/server/pm/PackageSetting;Lcom/android/internal/pm/parsing/pkg/ParsedPackage;Lcom/android/server/pm/Settings$VersionInfo;ZZZ)V
+    move/from16 v5, v16
+
+    iget-object v7, v15, Lcom/android/server/pm/InstallPackageHelper;->mCustomPlatformSignatures:[Landroid/content/pm/Signature;
+
+    invoke-static/range {v0 .. v7}, Lcom/android/server/pm/ScanPackageUtils;->collectCertificatesLI(Lcom/android/server/pm/PackageSetting;Lcom/android/internal/pm/parsing/pkg/ParsedPackage;Lcom/android/server/pm/pkg/AndroidPackage;Lcom/android/server/pm/Settings$VersionInfo;ZZZ[Landroid/content/pm/Signature;)V
+
+    move v3, v4
     :try_end_c
     .catch Lcom/android/server/pm/PackageManagerException; {:try_start_c .. :try_end_c} :catch_1
     .catchall {:try_start_c .. :try_end_c} :catchall_4
@@ -23995,7 +24051,9 @@
 
     iget-object v2, v2, Lcom/android/server/pm/PackageManagerService;->mPlatformPackage:Lcom/android/server/pm/pkg/AndroidPackage;
 
-    invoke-static {v1, v10, v2, v5}, Lcom/android/server/pm/ScanPackageUtils;->applyPolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;ILcom/android/server/pm/pkg/AndroidPackage;Z)V
+    iget-object v3, v15, Lcom/android/server/pm/InstallPackageHelper;->mCustomPlatformSignatures:[Landroid/content/pm/Signature;
+
+    invoke-static {v1, v10, v2, v5, v3}, Lcom/android/server/pm/ScanPackageUtils;->applyPolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;ILcom/android/server/pm/pkg/AndroidPackage;Z[Landroid/content/pm/Signature;)V
 
     iget-object v2, v15, Lcom/android/server/pm/InstallPackageHelper;->mPm:Lcom/android/server/pm/PackageManagerService;
 
diff --git a/smali_classes2/com/android/server/pm/PackageManagerServiceUtils.smali b/smali_classes2/com/android/server/pm/PackageManagerServiceUtils.smali
index c18a1a62..e20d5836 100644
--- a/smali_classes2/com/android/server/pm/PackageManagerServiceUtils.smali
+++ b/smali_classes2/com/android/server/pm/PackageManagerServiceUtils.smali
@@ -619,6 +619,16 @@
     return p0
 .end method
 
+.method public static compareSignaturesActual([Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)I
+    .locals 0
+
+    invoke-static {p0, p1}, Lcom/android/server/pm/PackageManagerServiceUtils;->compareSignatureArrays([Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)I
+
+    move-result p0
+
+    return p0
+.end method
+
 .method public static copyFile(Ljava/lang/String;Ljava/io/File;Ljava/lang/String;)V
     .locals 2
 
diff --git a/smali_classes2/com/android/server/pm/ScanPackageUtils.smali b/smali_classes2/com/android/server/pm/ScanPackageUtils.smali
index 62acf239..4d705051 100644
--- a/smali_classes2/com/android/server/pm/ScanPackageUtils.smali
+++ b/smali_classes2/com/android/server/pm/ScanPackageUtils.smali
@@ -159,7 +159,7 @@
     return-void
 .end method
 
-.method public static applyPolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;ILcom/android/server/pm/pkg/AndroidPackage;Z)V
+.method public static applyPolicy(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;ILcom/android/server/pm/pkg/AndroidPackage;Z[Landroid/content/pm/Signature;)V
     .locals 5
 
     const/high16 v0, 0x10000
@@ -367,9 +367,9 @@
 
     move-result p1
 
-    if-nez p1, :cond_b
+    if-nez p1, :cond_c
 
-    if-eqz p2, :cond_c
+    if-eqz p2, :cond_b
 
     invoke-interface {p2}, Lcom/android/server/pm/pkg/AndroidPackage;->getSigningDetails()Landroid/content/pm/SigningDetails;
 
@@ -383,15 +383,30 @@
 
     move-result p1
 
-    if-nez p1, :cond_c
+    if-eqz p1, :cond_c
 
     :cond_b
-    move v1, v2
+    invoke-interface {p0}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->getSigningDetails()Landroid/content/pm/SigningDetails;
+
+    move-result-object p1
+
+    invoke-virtual {p1}, Landroid/content/pm/SigningDetails;->getSignatures()[Landroid/content/pm/Signature;
+
+    move-result-object p1
+
+    invoke-static {p4, p1}, Lcom/android/server/pm/ScanPackageUtils;->signedWithCustomSignatures([Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)Z
+
+    move-result p1
+
+    if-eqz p1, :cond_d
 
     :cond_c
+    move v1, v2
+
+    :cond_d
     invoke-interface {p0, v1}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->setSignedWithPlatformKey(Z)Lcom/android/internal/pm/parsing/pkg/ParsedPackage;
 
-    if-nez v0, :cond_d
+    if-nez v0, :cond_e
 
     invoke-interface {p0}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->clearOriginalPackages()Lcom/android/internal/pm/parsing/pkg/ParsedPackage;
 
@@ -399,7 +414,7 @@
 
     invoke-interface {p1}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->clearAdoptPermissions()Lcom/android/internal/pm/parsing/pkg/ParsedPackage;
 
-    :cond_d
+    :cond_e
     invoke-static {p0, v0, p3}, Lcom/android/server/pm/parsing/library/PackageBackwardCompatibility;->modifySharedLibraries(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;ZZ)V
 
     return-void
@@ -882,20 +897,20 @@
     throw p0
 .end method
 
-.method public static collectCertificatesLI(Lcom/android/server/pm/PackageSetting;Lcom/android/internal/pm/parsing/pkg/ParsedPackage;Lcom/android/server/pm/Settings$VersionInfo;ZZZ)V
+.method public static collectCertificatesLI(Lcom/android/server/pm/PackageSetting;Lcom/android/internal/pm/parsing/pkg/ParsedPackage;Lcom/android/server/pm/pkg/AndroidPackage;Lcom/android/server/pm/Settings$VersionInfo;ZZZ[Landroid/content/pm/Signature;)V
     .locals 4
 
-    if-eqz p5, :cond_0
+    if-eqz p6, :cond_0
 
-    new-instance p5, Ljava/io/File;
+    new-instance p6, Ljava/io/File;
 
     invoke-interface {p1}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->getPath()Ljava/lang/String;
 
     move-result-object v0
 
-    invoke-direct {p5, v0}, Ljava/io/File;-><init>(Ljava/lang/String;)V
+    invoke-direct {p6, v0}, Ljava/io/File;-><init>(Ljava/lang/String;)V
 
-    invoke-virtual {p5}, Ljava/io/File;->lastModified()J
+    invoke-virtual {p6}, Ljava/io/File;->lastModified()J
 
     move-result-wide v0
 
@@ -907,11 +922,11 @@
     move-result-wide v0
 
     :goto_0
-    const-string/jumbo p5, "PackageManager"
+    const-string/jumbo p6, "PackageManager"
 
     if-eqz p0, :cond_4
 
-    if-nez p3, :cond_4
+    if-nez p4, :cond_4
 
     iget-object v2, p0, Lcom/android/server/pm/PackageSetting;->mPathString:Ljava/lang/String;
 
@@ -933,78 +948,88 @@
 
     sget-boolean v0, Lcom/android/server/pm/ReconcilePackageUtils;->ALLOW_NON_PRELOADS_SYSTEM_SHAREDUIDS:Z
 
-    iget p2, p2, Lcom/android/server/pm/Settings$VersionInfo;->databaseVersion:I
+    iget p3, p3, Lcom/android/server/pm/Settings$VersionInfo;->databaseVersion:I
 
     const/4 v0, 0x2
 
-    if-ge p2, v0, :cond_1
+    if-ge p3, v0, :cond_1
 
     goto :goto_1
 
     :cond_1
     const/4 v0, 0x3
 
-    if-ge p2, v0, :cond_2
+    if-ge p3, v0, :cond_2
 
     goto :goto_1
 
     :cond_2
-    iget-object p2, p0, Lcom/android/server/pm/PackageSetting;->signatures:Lcom/android/server/pm/PackageSignatures;
+    iget-object p3, p0, Lcom/android/server/pm/PackageSetting;->signatures:Lcom/android/server/pm/PackageSignatures;
 
-    iget-object p2, p2, Lcom/android/server/pm/PackageSignatures;->mSigningDetails:Landroid/content/pm/SigningDetails;
+    iget-object p3, p3, Lcom/android/server/pm/PackageSignatures;->mSigningDetails:Landroid/content/pm/SigningDetails;
 
-    invoke-virtual {p2}, Landroid/content/pm/SigningDetails;->getSignatures()[Landroid/content/pm/Signature;
+    invoke-virtual {p3}, Landroid/content/pm/SigningDetails;->getSignatures()[Landroid/content/pm/Signature;
 
-    move-result-object p2
+    move-result-object p3
 
-    if-eqz p2, :cond_3
+    if-eqz p3, :cond_3
 
-    iget-object p2, p0, Lcom/android/server/pm/PackageSetting;->signatures:Lcom/android/server/pm/PackageSignatures;
+    iget-object p3, p0, Lcom/android/server/pm/PackageSetting;->signatures:Lcom/android/server/pm/PackageSignatures;
 
-    iget-object p2, p2, Lcom/android/server/pm/PackageSignatures;->mSigningDetails:Landroid/content/pm/SigningDetails;
+    iget-object p3, p3, Lcom/android/server/pm/PackageSignatures;->mSigningDetails:Landroid/content/pm/SigningDetails;
 
-    invoke-virtual {p2}, Landroid/content/pm/SigningDetails;->getSignatures()[Landroid/content/pm/Signature;
+    invoke-virtual {p3}, Landroid/content/pm/SigningDetails;->getSignatures()[Landroid/content/pm/Signature;
 
-    move-result-object p2
+    move-result-object p3
 
-    array-length p2, p2
+    array-length p3, p3
 
-    if-eqz p2, :cond_3
+    if-eqz p3, :cond_3
 
-    iget-object p2, p0, Lcom/android/server/pm/PackageSetting;->signatures:Lcom/android/server/pm/PackageSignatures;
+    iget-object p3, p0, Lcom/android/server/pm/PackageSetting;->signatures:Lcom/android/server/pm/PackageSignatures;
 
-    iget-object p2, p2, Lcom/android/server/pm/PackageSignatures;->mSigningDetails:Landroid/content/pm/SigningDetails;
+    iget-object p3, p3, Lcom/android/server/pm/PackageSignatures;->mSigningDetails:Landroid/content/pm/SigningDetails;
 
-    invoke-virtual {p2}, Landroid/content/pm/SigningDetails;->getSignatureSchemeVersion()I
+    invoke-virtual {p3}, Landroid/content/pm/SigningDetails;->getSignatureSchemeVersion()I
 
-    move-result p2
+    move-result p3
+
+    if-eqz p3, :cond_3
 
-    if-eqz p2, :cond_3
+    new-instance p3, Landroid/content/pm/SigningDetails;
 
-    new-instance p2, Landroid/content/pm/SigningDetails;
+    iget-object v0, p0, Lcom/android/server/pm/PackageSetting;->signatures:Lcom/android/server/pm/PackageSignatures;
+
+    iget-object v0, v0, Lcom/android/server/pm/PackageSignatures;->mSigningDetails:Landroid/content/pm/SigningDetails;
+
+    invoke-direct {p3, v0}, Landroid/content/pm/SigningDetails;-><init>(Landroid/content/pm/SigningDetails;)V
+
+    invoke-interface {p1, p3}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->setSigningDetails(Landroid/content/pm/SigningDetails;)Lcom/android/internal/pm/parsing/pkg/ParsedPackage;
 
     iget-object p0, p0, Lcom/android/server/pm/PackageSetting;->signatures:Lcom/android/server/pm/PackageSignatures;
 
     iget-object p0, p0, Lcom/android/server/pm/PackageSignatures;->mSigningDetails:Landroid/content/pm/SigningDetails;
 
-    invoke-direct {p2, p0}, Landroid/content/pm/SigningDetails;-><init>(Landroid/content/pm/SigningDetails;)V
+    invoke-virtual {p0}, Landroid/content/pm/SigningDetails;->getSignatures()[Landroid/content/pm/Signature;
+
+    move-result-object p0
 
-    invoke-interface {p1, p2}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->setSigningDetails(Landroid/content/pm/SigningDetails;)Lcom/android/internal/pm/parsing/pkg/ParsedPackage;
+    invoke-static {p1, p2, p7, p0}, Lcom/android/server/pm/ScanPackageUtils;->setCustomSignatures(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;Lcom/android/server/pm/pkg/AndroidPackage;[Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)V
 
     return-void
 
     :cond_3
-    new-instance p2, Ljava/lang/StringBuilder;
+    new-instance p3, Ljava/lang/StringBuilder;
 
-    const-string/jumbo p3, "PackageSetting for "
+    const-string/jumbo p4, "PackageSetting for "
 
-    invoke-direct {p2, p3}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
+    invoke-direct {p3, p4}, Ljava/lang/StringBuilder;-><init>(Ljava/lang/String;)V
 
     iget-object p0, p0, Lcom/android/server/pm/PackageSetting;->mName:Ljava/lang/String;
 
-    const-string p3, " is missing signatures.  Collecting certs again to recover them."
+    const-string p4, " is missing signatures.  Collecting certs again to recover them."
 
-    invoke-static {p2, p0, p3, p5}, Lcom/android/server/ProfileService$$ExternalSyntheticOutline0;->m(Ljava/lang/StringBuilder;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)V
+    invoke-static {p3, p0, p4, p6}, Lcom/android/server/ProfileService$$ExternalSyntheticOutline0;->m(Ljava/lang/StringBuilder;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)V
 
     goto :goto_3
 
@@ -1016,39 +1041,39 @@
 
     invoke-interface {p1}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->getPath()Ljava/lang/String;
 
-    move-result-object p2
+    move-result-object p3
 
-    invoke-virtual {p0, p2}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+    invoke-virtual {p0, p3}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
 
-    const-string p2, " changed; collecting certs"
+    const-string p3, " changed; collecting certs"
 
-    invoke-virtual {p0, p2}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+    invoke-virtual {p0, p3}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
 
-    if-eqz p3, :cond_5
+    if-eqz p4, :cond_5
 
-    const-string p2, " (forced)"
+    const-string p3, " (forced)"
 
     goto :goto_2
 
     :cond_5
-    const-string p2, ""
+    const-string p3, ""
 
     :goto_2
-    invoke-static {p0, p2, p5}, Lcom/android/server/DeviceIdleController$$ExternalSyntheticOutline0;->m(Ljava/lang/StringBuilder;Ljava/lang/String;Ljava/lang/String;)V
+    invoke-static {p0, p3, p6}, Lcom/android/server/DeviceIdleController$$ExternalSyntheticOutline0;->m(Ljava/lang/StringBuilder;Ljava/lang/String;Ljava/lang/String;)V
 
     :goto_3
-    const-wide/32 p2, 0x40000
+    const-wide/32 v2, 0x40000
 
     :try_start_0
     const-string/jumbo p0, "collectCertificates"
 
-    invoke-static {p2, p3, p0}, Landroid/os/Trace;->traceBegin(JLjava/lang/String;)V
+    invoke-static {v2, v3, p0}, Landroid/os/Trace;->traceBegin(JLjava/lang/String;)V
 
     invoke-static {}, Landroid/content/pm/parsing/result/ParseTypeImpl;->forDefaultParsing()Landroid/content/pm/parsing/result/ParseTypeImpl;
 
     move-result-object p0
 
-    invoke-static {p0, p1, p4}, Lcom/android/internal/pm/pkg/parsing/ParsingPackageUtils;->getSigningDetails(Landroid/content/pm/parsing/result/ParseInput;Lcom/android/internal/pm/parsing/pkg/ParsedPackage;Z)Landroid/content/pm/parsing/result/ParseResult;
+    invoke-static {p0, p1, p5}, Lcom/android/internal/pm/pkg/parsing/ParsingPackageUtils;->getSigningDetails(Landroid/content/pm/parsing/result/ParseInput;Lcom/android/internal/pm/parsing/pkg/ParsedPackage;Z)Landroid/content/pm/parsing/result/ParseResult;
 
     move-result-object p0
 
@@ -1065,10 +1090,16 @@
     check-cast p0, Landroid/content/pm/SigningDetails;
 
     invoke-interface {p1, p0}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->setSigningDetails(Landroid/content/pm/SigningDetails;)Lcom/android/internal/pm/parsing/pkg/ParsedPackage;
+
+    invoke-virtual {p0}, Landroid/content/pm/SigningDetails;->getSignatures()[Landroid/content/pm/Signature;
+
+    move-result-object p3
+
+    invoke-static {p1, p2, p7, p3}, Lcom/android/server/pm/ScanPackageUtils;->setCustomSignatures(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;Lcom/android/server/pm/pkg/AndroidPackage;[Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)V
     :try_end_0
     .catchall {:try_start_0 .. :try_end_0} :catchall_0
 
-    invoke-static {p2, p3}, Landroid/os/Trace;->traceEnd(J)V
+    invoke-static {v2, v3}, Landroid/os/Trace;->traceEnd(J)V
 
     return-void
 
@@ -1100,7 +1131,7 @@
     .catchall {:try_start_1 .. :try_end_1} :catchall_0
 
     :goto_4
-    invoke-static {p2, p3}, Landroid/os/Trace;->traceEnd(J)V
+    invoke-static {v2, v3}, Landroid/os/Trace;->traceEnd(J)V
 
     throw p0
 .end method
@@ -4871,6 +4902,31 @@
     throw v0
 .end method
 
+.method public static setCustomSignatures(Lcom/android/internal/pm/parsing/pkg/ParsedPackage;Lcom/android/server/pm/pkg/AndroidPackage;[Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)V
+    .locals 0
+
+    invoke-static {p2, p3}, Lcom/android/server/pm/ScanPackageUtils;->signedWithCustomSignatures([Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)Z
+
+    move-result p2
+
+    if-eqz p2, :cond_1
+
+    if-nez p1, :cond_0
+
+    goto :goto_0
+
+    :cond_0
+    invoke-interface {p1}, Lcom/android/server/pm/pkg/AndroidPackage;->getSigningDetails()Landroid/content/pm/SigningDetails;
+
+    move-result-object p1
+
+    invoke-interface {p0, p1}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->setSigningDetails(Landroid/content/pm/SigningDetails;)Lcom/android/internal/pm/parsing/pkg/ParsedPackage;
+
+    :cond_1
+    :goto_0
+    return-void
+.end method
+
 .method public static setInstantAppForUser(Lcom/android/server/pm/PackageManagerServiceInjector;Lcom/android/server/pm/PackageSetting;IZZ)V
     .locals 6
 
@@ -5003,3 +5059,23 @@
     :goto_2
     return-void
 .end method
+
+.method public static signedWithCustomSignatures([Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)Z
+    .locals 0
+
+    invoke-static {p0, p1}, Lcom/android/server/pm/PackageManagerServiceUtils;->compareSignaturesActual([Landroid/content/pm/Signature;[Landroid/content/pm/Signature;)I
+
+    move-result p0
+
+    if-nez p0, :cond_0
+
+    const/4 p0, 0x1
+
+    goto :goto_0
+
+    :cond_0
+    const/4 p0, 0x0
+
+    :goto_0
+    return p0
+.end method
-- 
2.51.0
